File: go/39.go
/*

https://leetcode.com/problems/combination-sum

Given a set of candidate numbers (candidates) (without duplicates) and a target number (target), find all unique combinations in candidates where the candidate numbers sums to target.

The same repeated number may be chosen from candidates unlimited number of times.

Note:

All numbers (including target) will be positive integers.
The solution set must not contain duplicate combinations.
Example 1:

Input: candidates = [2,3,6,7], target = 7,
A solution set is:
[
  [7],
  [2,2,3]
]
Example 2:

Input: candidates = [2,3,5], target = 8,
A solution set is:
[
  [2,2,2,2],
  [2,3,3],
  [3,5]
]
*/


package main


import "fmt"
import "sort"


func combinationSum(candidates []int, target int) [][]int {
    sort.Ints(candidates)
    res := make([][]int, 0)
    var f func(buf []int, n int)
    f = func(buf []int, n int) {
        if n == 0 {
            tmp := make([]int, len(buf))
            copy(tmp, buf)
            res = append(res, tmp)
            return
        }
        for _, c := range candidates {
            if c > n {
                break
            }        
            // cut off
            // 非常经典的剪枝
            if len(buf) > 0 && buf[len(buf) - 1] > c {
                continue
            }
            buf = append(buf, c)
            f(buf, n - c)
            buf = buf[:len(buf) - 1]
        }
    }
    buf := make([]int, 0)
    f(buf, target)
    return res
}


func main() {
    candidates := []int{2,3,6,7}
    target := 7
    fmt.Println(combinationSum(candidates, target))
}

File: python/39.py
"""
https://leetcode.com/problems/combination-sum

Input: candidates = [2,3,6,7], target = 7,
A solution set is:
[
  [7],
  [2,2,3]
]
Example 2:

Input: candidates = [2,3,5], target = 8,
A solution set is:
[
  [2,2,2,2],
  [2,3,3],
  [3,5]
]

class Solution(object):
def combinationSum(self, candidates, target):
    result = []
    candidates = sorted(candidates)
    def dfs(remain, stack):
        if remain == 0:
            result.append(stack)
            return 

        for item in candidates:
            if item > remain: break
            if stack and item < stack[-1]: continue
            else:
                dfs(remain - item, stack + [item])
    
    dfs(target, [])
    return result
"""

from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()

        res = set()

        def f(buf, t):
            # print(buf, t)
            if t == 0:
                res.add(tuple(sorted(buf[:])))
                return
            for c in candidates:
                if c > t:
                    break
                tmp = buf[:]
                tmp.append(c)
                f(tmp, t - c)
                # f(buf, t)

        f([], target)
        return [list(i) for i in res]

# 0 - 1 背包

class Solution2:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = set()
        
        def f(buf, t):
            if t == 0:
                res.add(tuple(sorted(buf[:])))
                return
            for c in candidates:
                if c > t:
                    break
                buf.append(c)
                f(buf[:], t - c)
                buf.pop()

        f([], target)
        return [list(i) for i in res]



class Solution3:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        
        def f(buf, t):
            if t == 0:
                res.append(buf[:])
                return
            for c in candidates:
                if c > t:
                    break
                if buf and c < buf[-1]:
                    print("hit", buf, c, t)
                    continue
                buf.append(c)
                f(buf[:], t - c)
                buf.pop()

        f([], target)
        return res


class Solution4:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        
        def f(buf, t):
            if t == 0:
                res.append(buf[:])
                return

            for c in candidates:
                if c > t:
                    break
                if buf and c < buf[-1]:
                    # print("hit", buf, c, t)
                    continue
                buf.append(c)
                f(buf, t - c)
                buf.pop()

        f([], target)
        return res


if __name__ == "__main__":
    cases = [
        ([2, 3, 6, 7], 7),
        ([2, 3, 5], 8),
    ]
    for candidates, target in cases:
        print(Solution4().combinationSum(candidates, target))
